core/arch/arm/kernel/boot.c

```diff
@@ -692,32 +692,59 @@ static int add_dt_path_subnode(struct dt_descriptor *dt, const char *path,
 
 static int add_optee_dt_node(struct dt_descriptor *dt)
 {
-	int offs;
-	int ret;
+	int offs = 0;
+	int ret = 0;
 
-	if (fdt_path_offset(dt->blob, "/firmware/optee") >= 0) {
-		DMSG("OP-TEE Device Tree node already exists!");
-		return 0;
-	}
+	if (IS_ENABLED(CFG_CORE_SEL1_SPMC)) {
+		const char *uuid = "486178e0-e7f8-11e3-bc5e-0002a5d5c51b";
 
-	offs = fdt_path_offset(dt->blob, "/firmware");
-	if (offs < 0) {
-		offs = add_dt_path_subnode(dt, "/", "firmware");
+		if (fdt_path_offset(dt->blob, "/ffa/optee") >= 0) {
+			DMSG("OP-TEE Device Tree node already exists!");
+			return 0;
+		}
+
+		offs = fdt_path_offset(dt->blob, "/ffa");
+		if (offs < 0) {
+			EMSG("FF-A node \"/ffa\" missing!");
+			return -1;
+		}
+
+		offs = fdt_add_subnode(dt->blob, offs, "optee");
```

Comment:
Where is the DT binding defined? As I understand from [this LKML discussion](https://patchwork.kernel.org/project/linux-arm-kernel/patch/20210325143255.1532452-2-sudeep.holla@arm.com/) this might not even be needed?

Reply:
FF-A is still evolving. It is very likely that we can remove this later, but currently it's needed for this prototype to work.

So we have two options here:
1. have a complete SPMC at S-EL1 prototype as we've had until now
2. keep a delta in separate fork and rebase from time to time hoping that conflicts will not be too hard

My strategy with the upstream FF-A support in OP-TEE is that it's experimental as long as the counter parts in Linux are not merged yet. While FF-A support is still experimental we can make changes in ABIs etc. So while this may lead to merging stuff upstream which isn't final it's very convenient as it avoids many conflicts by upstreaming early.

Reply:
I understand. My concern is, even marked "experimental" people may start using the binding since there is nothing else, and then we could be hesitant when it's time to deprecate it (especially if this stays experimental for a long time which it honestly seems it will given the number of components/projects involved in the architecture). Rob's answer was pretty clear, make things discoverable if you can.

Reply:
There's not much I can do about that. I'll keep this part in the fork until there's a decision.